src/java/org/apache/cassandra/io/util/ThreadLocalReadAheadBuffer.java

```diff
@@ -0,0 +1,154 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.util;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.netty.util.concurrent.FastThreadLocal;
+import org.apache.cassandra.io.compress.BufferType;
+import org.apache.cassandra.io.compress.CorruptBlockException;
+import org.apache.cassandra.io.sstable.CorruptSSTableException;
+
+public final class ThreadLocalReadAheadBuffer
+{
+
+    private final ChannelProxy channel;
+
+    private final BufferType bufferType;
+
+    private static final Map<String, FastThreadLocal<ByteBuffer>> blockBufferHolders = new ConcurrentHashMap<>();
+
+    private static final Map<String, FastThreadLocal<Integer>> storedBlockNos = new ConcurrentHashMap<>();
+
+    private final FastThreadLocal<ByteBuffer> blockBufferHolder;
+
+    private final FastThreadLocal<Integer> storedBlockNo;
+
+    private final int bufferSize;
+
+    public ThreadLocalReadAheadBuffer(ChannelProxy channel, int bufferSize, BufferType bufferType)
+    {
+        this.channel = channel;
+        this.bufferSize = bufferSize;
+        this.bufferType = bufferType;
+        blockBufferHolders.putIfAbsent(channel.filePath(), new FastThreadLocal<>()
+        {
+            @Override
+            protected ByteBuffer initialValue() throws Exception
+            {
+                return null;
+            }
+        });
+        storedBlockNos.putIfAbsent(channel.filePath(), new FastThreadLocal<>() {
+            @Override
+            protected Integer initialValue() throws Exception
+            {
+                return -1;
+            }
+        });
+        blockBufferHolder = blockBufferHolders.get(channel.filePath());
+        storedBlockNo = storedBlockNos.get(channel.filePath());
+    }
+
+    public boolean hasBuffer()
+    {
+        return blockBufferHolder.get() != null;
+    }
+
+    public int remaining()
+    {
+        return getBlockBuffer().remaining();
+    }
+
+
+    public ByteBuffer getBlockBuffer()
+    {
+        ByteBuffer buffer = blockBufferHolder.get();
+        if (buffer == null)
+        {
+            buffer = bufferType.allocate(bufferSize);
+            buffer.clear();
+            blockBufferHolder.set(buffer);
+        }
+        return buffer;
+    }
+
+    public void fill(long position) throws CorruptBlockException
+    {
+        ByteBuffer blockBuffer = getBlockBuffer();
+        long channelSize = channel.size();
+        long realPosition = Math.min(channelSize, position);
+        long blockLength = blockBuffer.limit();
+        int blockNo = (int) (realPosition / blockLength);
+        long blockPosition = blockNo * blockLength;
+
+        if (storedBlockNo.get() != blockNo)
```

Comment:
nit: Quick comment on this being here/necessary to skip when we've already read 256KiB worth of 16KiB chunks and just need to advance `blockBuffer`